### PR TITLE
handle upcoming asdf deprecations

### PR DESCRIPTION
**Pull request by braingram**
asdf is deprecating AsdfInFits and resolvers. These are both used in roman_datamodels.
https://github.com/asdf-format/asdf/pull/1337
https://github.com/asdf-format/asdf/pull/1362

This PR conditionally handles importing AsdfInFits (based on the asdf version) for checking files to raise an exception when code attempts to use AsdfInFits for a datamodel.

Resolvers are older versions of resource mappings. The uses of resolver in this package can be removed as the rad schema are registered [using a resource mapping, not a resolver](https://github.com/spacetelescope/rad/blob/5b64a95600e0130800a1c13a710c046c2c361a1f/pyproject.toml#L43)) and the default resolver argument to ` asdf.schema.get_validator` is sufficient.

### CODE REVIEW

**Review by perrygreenfield — APPROVED** [review]
LGTM

**Comment by braingram** [comment]
Testing locally with the current development version of rad. I see the same test failures I see when testing main (with rad main):
```
============================================================================================================== short test summary info ==============================================================================================================
FAILED tests/test_factories.py::test_factory_method_implemented[Associations] - ValueError: Factory method not implemented for class Associations
FAILED tests/test_factories.py::test_instance_valid[Associations] - ValueError: Factory method not implemented for class Associations
FAILED tests/test_stnode.py::test_serialization[Associations] - ValueError: Factory method not implemented for class Associations
```

For reference here was a PR that switched main to testing against rad and showed the same errors. https://github.com/spacetelescope/roman_datamodels/pull/127#issuecomment-1448482619

**Comment by braingram** [comment]
It appears I do not have write access to this repository and cannot merge this PR. @perrygreenfield do you think we should merge this now or wait till the CI is repaired?

**Comment by braingram** [comment]
@PaulHuwe @ddavis-stsci should regression tests be run for romancal with these changes? If so, is there documentation on how to run these? I attempted to follow a similar procedure as described for the jwst pipeline:
https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches
but do not appear to have jenkins permissions to configure the `Pipeline Roman-Developers-Pull-Requests` job. When I attempted to start it with the override parameter to install this branch (`git+https://github.com/braingram/roman_datamodels.git@asdf_deprecations`) it immediately failed with
```
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
ERROR: Maximum checkout retry attempts reached, aborting
```
https://plwishmaster.stsci.edu:8081/job/RT/job/Roman-Developers-Pull-Requests/97/console

File: tests/test_open.py
Comment (braingram) [reply tests/test_open.py]: Thanks for the catch. It looks like `3.dev` will also work:
```
In [13]: Version("3.0.0.dev6+g5b64a95") >= Version("3.dev")
Out[13]: True
```
I switched the comparisons to using that.

**Comment by braingram** [comment]
To provide an update, regression tests of this PR are held up by the `roman_datamodels` requirement specified in romancal which pulls from the `spacetelescope/roman_datamodels` repo directly:
https://github.com/spacetelescope/romancal/blob/96f919b85546c5c22706b6bed2ad57de14a5339e/pyproject.toml#L25
This prevents the overrides in the plwishmaster job from correctly pulling from a non-spacetelescope fork (`braingram/roman_datamodels` in this case). Until a work around for this, or a change to romancal devs are made, regression tests on this PR are not possible.

**Comment by braingram** [comment]
Regression tests run and passed here: https://plwishmaster.stsci.edu:8081/job/RT/job/Roman-Developers-Pull-Requests/108/

**Pull request merged**
